### PR TITLE
Run integration tests on the CI for the PHP client of the API

### PR DESCRIPTION
**Pull request by ahocquard**
Branch merged in this PR: https://github.com/akeneo/php-api-client/pull/39 (CI bug => have to create a new PR)

**The goal of this PR is to run integration tests on the CI for the PHP client of the API.**

**Important note**: the second commit with tests is just to test my PR and will be deleted before merging my PR.

Testing the API with integration tests implies to reset the database between each test.

This is not an easy task, because the PIM is in a docker and the client in another one.
The two tasks our PHPUnit should do are:
- reset the database
- generate the new client_id/secret to test the API

To be able to perform that, there is two possible configurations:
- one for local, performing `pim:installer:db` and `pim:oauth-server:create-client` directly in local
- one for docker, performing these commands thanks to `docker exec` commands

The second one has been implemented in this PR, by sharing the docker socket and the docker bin to the container. It works pretty nice and allows to communicate inside a container with some other containers.

We use a virtual package for the HTTP client of our client. It allows to use any client implementing the virtual package `php-http/client-implementation`(https://packagist.org/providers/php-http/client-implementation).

Moreover, 3 implementations of PSR7 could be used with those packages:
- guzzlehttp/psr7
- zendframework/zend-diactoros
- slim/slim

So, we have a matrice [php version * HTTP client * Psr implementations ] to test if we want to have a good coverage with our client.

If it's too costly to test everything in the future, we can easily deactivate a client, a version, a PSR7 implem.

For the moment, we are just testing the most used HTTP client, and will adapt it according to our needs.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -28,15 +34,43 @@ stage("Checkout") {
         deleteDir()
         checkout scm
         stash "php-api-client"
+
+       checkout([$class: 'GitSCM',
+         branches: [[name: '1.7']],
```
Comment (ClementGautier) [inline Jenkinsfile]: pimVersion here I guess

File: Jenkinsfile [hunk 2]
```diff
@@ -64,29 +101,71 @@ if (launchIntegrationTests.equals("yes")) {
...
+
+/**
+ * Run checkout of the PHP client, for a given PHP version, HTTP client and PSR7 implementation.
+ *
+ * @param phpVersion PHP version to use to run the composer
+ * @param client     name of the HTTP client package to use to checkout
+ * @param psrImplem  name of the PSR 7 implementation package to checkout
+ */
+void runCheckoutClient(String phpVersion, String client, String psrImplem) {
     node('docker') {
```
Comment (ClementGautier) [inline Jenkinsfile]: I would add a check here: if the stash already exist you can do an early return

File: Jenkinsfile [hunk 3]
```diff
@@ -64,29 +101,71 @@ if (launchIntegrationTests.equals("yes")) {
...
 
-def runCheckout(phpVersion, client) {
+/**
+ * Run checkout of the PIM for a given PHP version and a PIM version.
+ *
+ * @param phpVersion PHP version to use to run the composer
+ * @param pimVersion PIM version to checkout
+ */
+void runCheckoutPim(String phpVersion, String pimVersion) {
+    node('docker') {
```
Comment (ClementGautier) [inline Jenkinsfile]: I would add a check here: if the stash already exist you can do an early return

File: Jenkinsfile [hunk 4]
```diff
@@ -64,29 +101,71 @@ if (launchIntegrationTests.equals("yes")) {
...
+        deleteDir()
+        try {
+            docker.image("carcel/php:${phpVersion}").inside("-v /home/akeneo/.composer:/home/docker/.composer") {
+                unstash "pim_community_dev_${pimVersion}"
```
Comment (ClementGautier) [inline Jenkinsfile]: what if this stash doesn't exist ? You should probably do the checkout here

File: Jenkinsfile [hunk 5]
```diff
@@ -47,12 +81,15 @@ if (launchUnitTests.equals("yes")) {
     stage("Unit tests and Code style") {
         def tasks = [:]
 
-        tasks["php-cs-fixer"] = {runPhpCsFixerTest()}
+        String currentClient = "php-http/guzzle6-adapter"
+        String currentPsrImplem = "guzzlehttp/psr7"
 
-        for (phpVersion in supportedPhpVersions) {
-            def currentVersion = phpVersion
+        tasks["php-cs-fixer"] = {runPhpCsFixerTest("7.1", currentClient, currentPsrImplem)}
```
Comment (ClementGautier) [inline Jenkinsfile]: phpcs rules can be different from one version to another, so, personaly, I'd put it for each supported version but it's as you which

File: Jenkinsfile [hunk 6]
```diff
@@ -1,24 +1,30 @@
...
 
-def launchUnitTests = "yes"
-def launchIntegrationTests = "no"
-def clients = ["php-http/guzzle5-adapter", "php-http/guzzle6-adapter"]
-// def pimVersion = "1.7"
+def clientConfig = [
+    "php-http/guzzle6-adapter": ["phpVersion": ["5.6", "7.0", "7.1"], "psrImplem": ["guzzlehttp/psr7"]],
+    "php-http/guzzle5-adapter": ["phpVersion": ["5.6", "7.0", "7.1"], "psrImplem": ["guzzlehttp/psr7", "zendframework/zend-diactoros", "slim/slim"]],
+    "php-http/curl-client": ["phpVersion": ["5.6", "7.0", "7.1"], "psrImplem": ["guzzlehttp/psr7", "zendframework/zend-diactoros", "slim/slim"]]
+    //"php-http/socket-client": ["phpVersion": ["5.6", "7.0", "7.1"], "psrImplem": ["guzzlehttp/psr7", "zendframework/zend-diactoros", "slim/slim"]]
```
Comment (ClementGautier) [inline Jenkinsfile]: don't merge commented code :P it's a debt you will forget soon, better remove it

File: Jenkinsfile (hunk 5 above)
Comment (ahocquard) [inline Jenkinsfile]: Following the rules for 7.1 is enough.

File: Jenkinsfile [hunk 7]
```diff
@@ -28,15 +34,43 @@ stage("Checkout") {
         deleteDir()
         checkout scm
         stash "php-api-client"
+
```
Comment (ahocquard) [inline Jenkinsfile]: Done like that in the PIM, not needed

File: Jenkinsfile (hunk 3 above)
Comment (ahocquard) [inline Jenkinsfile]: Why ? It can't happen actually

File: Jenkinsfile (hunk 4 above)
Comment (ahocquard) [inline Jenkinsfile]: same : can't happen.

File: Jenkinsfile (hunk 2 above)
Comment (ahocquard) [inline Jenkinsfile]: Same ;)

File: Jenkinsfile [hunk 8]
```diff
@@ -64,29 +98,71 @@ if (launchIntegrationTests.equals("yes")) {
...
+ * @param pimVersion PIM version to checkout
+ */
+void runCheckoutPim(String phpVersion, String pimVersion) {
+    node('docker') {
+        deleteDir()
+        try {
+            docker.image("carcel/php:${phpVersion}").inside("-v /home/akeneo/.composer:/home/docker/.composer") {
+                unstash "pim_community_dev_${pimVersion}"
+
+                sh "composer require \"akeneo/catalogs\":\"dev-API-175\" --ignore-platform-reqs --optimize-autoloader --no-interaction --no-progress --prefer-dist"
```
Comment (skeleton) [inline Jenkinsfile]: merge API-175 and change branch name before merge please

File: tests/DockerCredentialGenerator.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        $this->dockerName = $dockerName;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generate($path)
+    {
+        $command = sprintf('docker exec %s php %s/app/console pim:oauth-server:create-client -e prod', $this->dockerName, $path);
```
Comment (skeleton) [inline tests/DockerCredentialGenerator.php]: why prod ? we are in test env

File: Jenkinsfile [hunk 10]
```diff
@@ -1,42 +1,73 @@
...
-            // choice(choices: '1.7', description: 'PIM version to run integration tests with', name: 'pimVersion'),
+            choice(choices: 'yes\nno', description: 'Run integration tests', name: 'launchIntegrationTests'),
+            string(defaultValue: clients.join(','), description: 'Clients used to run integration tests (comma separated values)', name: 'clients'),
         ])
 
         launchUnitTests = userInput['launchUnitTests']
-        // launchIntegrationTests = userInput['launchIntegrationTests']
+        launchIntegrationTests = userInput['launchIntegrationTests']
         clients = userInput['clients'].tokenize(',')
-        // pimVersion = userInput['pimVersion']
```
Comment (skeleton) [inline Jenkinsfile]: remove comments please

File: Jenkinsfile [hunk 11]
```diff
@@ -1,42 +1,73 @@
...
     milestone 2
 
     node {
         deleteDir()
         checkout scm
         stash "php-api-client"
+
+       checkout([$class: 'GitSCM',
+         branches: [[name: pimVersion]],
+         userRemoteConfigs: [[credentialsId: 'github-credentials', url: 'https://github.com/akeneo/pim-community-dev.git']]
```
Comment (skeleton) [inline Jenkinsfile]: we will talk later about EE repo

Comment (ahocquard) [inline Jenkinsfile]: yep, and 1.8 . But I wanted to have a working solution at first.

File: tests/DockerCredentialGenerator.php (hunk 9 above)
Comment (ahocquard) [inline tests/DockerCredentialGenerator.php]: On the client side, yes, but useless on server side.

File: Jenkinsfile [hunk 12]
```diff
@@ -1,42 +1,73 @@
...
+String launchUnitTests = "yes"
+String launchIntegrationTests = "yes"
+String pimVersion = "1.7"
 
-def launchUnitTests = "yes"
-def launchIntegrationTests = "no"
-def clients = ["php-http/guzzle5-adapter", "php-http/guzzle6-adapter"]
-// def pimVersion = "1.7"
+def clientConfig = [
+    "php-http/guzzle6-adapter": ["phpVersion": ["5.6", "7.0", "7.1"], "psrImplem": ["guzzlehttp/psr7"]],
```
Comment (damien-carcel) [inline Jenkinsfile]: Why not using `supportedPhpVersions` here (and below)? You would have only one line to change when you'll had PHP 7.2.

File: Jenkinsfile (hunk 7 above)
Comment (damien-carcel) [inline Jenkinsfile]: Mh, Clément has a point. Maybe it is the PIM Jenkinsfile that is wrong.

File: Jenkinsfile (hunk 5 above)
Comment (damien-carcel) [inline Jenkinsfile]: I couldn't find anything in`php-cs-fixer` documentation stating that the rules change according to PHP version. I would think it has no impact at all.

File: Jenkinsfile (hunk 12 above)
Comment (ahocquard) [inline Jenkinsfile]: At first, it was because the supported versions could be different between the HTTP clients.

But no usage for now: I will fix it.

**Pull request closed without merging**